File: src/api/knowledge.ts
import type { KnowledgeBase, KnowledgeDoc, SearchResult } from "@/types/knowledge";
import { isTauri } from "./env";
import { mockKnowledgeBases, mockKnowledgeDocs } from "@/mocks/data";

function genId(prefix: string) {
  return `${prefix}_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
}

export async function listKnowledgeBases(): Promise<KnowledgeBase[]> {
  if (!isTauri()) return structuredClone(mockKnowledgeBases);
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<KnowledgeBase[]>("list_knowledge_bases");
}

export async function createKnowledgeBase(name: string, description?: string): Promise<KnowledgeBase> {
  if (!isTauri()) {
    const kb: KnowledgeBase = {
      id: genId("kb"),
      name,
      description: description ?? "",
      docCount: 0,
    };
    mockKnowledgeBases.push(kb);
    return { ...kb };
  }
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<KnowledgeBase>("create_knowledge_base", { name, description: description ?? null });
}

export async function listKnowledgeDocs(kbId: string): Promise<KnowledgeDoc[]> {
  if (!isTauri()) {
    return structuredClone(mockKnowledgeDocs).filter((d) => d.kbName === kbId);
  }
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<KnowledgeDoc[]>("list_knowledge_docs", { kbId });
}

export async function importDocument(
  kbId: string,
  title: string,
  source: string,
  content: string,
  chunkType?: string,
): Promise<string> {
  if (!isTauri()) {
    const doc: KnowledgeDoc = {
      id: genId("doc"),
      kbName: kbId,
      title,
      source,
      chunkType: (chunkType as KnowledgeDoc["chunkType"]) ?? "text",
      chunkCount: Math.max(1, Math.ceil(content.length / 500)),
      createdAt: new Date().toISOString(),
    };
    mockKnowledgeDocs.push(doc);
    const kb = mockKnowledgeBases.find((k) => k.id === kbId || k.name === kbId);
    if (kb) kb.docCount += 1;
    return doc.id;
  }
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<string>("import_document", {
    kbId,
    title,
    source,
    content,
    chunkType: chunkType ?? null,
  });
}

export async function deleteDoc(docId: string): Promise<void> {
  if (!isTauri()) {
    const idx = mockKnowledgeDocs.findIndex((d) => d.id === docId);
    if (idx >= 0) {
      const doc = mockKnowledgeDocs[idx];
      mockKnowledgeDocs.splice(idx, 1);
      const kb = mockKnowledgeBases.find((k) => k.name === doc.kbName);
      if (kb && kb.docCount > 0) kb.docCount -= 1;
    }
    return;
  }
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke("delete_doc", { docId });
}

export async function searchKnowledge(
  query: string,
  kbId?: string,
  topK?: number,
): Promise<SearchResult[]> {
  if (!isTauri()) {
    const q = query.toLowerCase();
    const docs = mockKnowledgeDocs.filter((d) => !kbId || d.kbName === kbId);
    return docs
      .map((d) => ({
        chunkId: genId("chunk"),
        text: `[${d.title}] ${d.source}`,
        score: Math.random() * 0.5 + 0.3,
        docTitle: d.title,
      }))
      .filter((r) => r.text.toLowerCase().includes(q))
      .slice(0, topK ?? 5);
  }
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<SearchResult[]>("search_knowledge", {
    query,
    kbId: kbId ?? null,
    topK: topK ?? null,
  });
}

File: src-tauri/src/commands/knowledge.rs
use tauri::State;
use crate::state::AppState;

#[tauri::command]
pub fn list_knowledge_bases(state: State<'_, AppState>) -> Result<Vec<serde_json::Value>, String> {
    state.db.with_conn(|conn| {
        let mut stmt = conn.prepare("SELECT id, name, description, created_at FROM knowledge_base ORDER BY name")?;
        let rows = stmt.query_map([], |row| {
            Ok(serde_json::json!({
                "id": row.get::<_, String>(0)?,
                "name": row.get::<_, String>(1)?,
                "description": row.get::<_, Option<String>>(2)?,
                "createdAt": row.get::<_, String>(3)?,
            }))
        })?;
        rows.collect::<Result<Vec<_>, _>>()
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_knowledge_base(state: State<'_, AppState>, name: String, description: Option<String>) -> Result<serde_json::Value, String> {
    let id = format!("kb_{}", uuid::Uuid::new_v4());
    state.db.with_conn(|conn| {
        conn.execute(
            "INSERT INTO knowledge_base (id, name, description) VALUES (?1, ?2, ?3)",
            rusqlite::params![id, name, description],
        )?;
        Ok(serde_json::json!({"id": id, "name": name, "description": description}))
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn list_knowledge_docs(state: State<'_, AppState>, kb_id: String) -> Result<Vec<serde_json::Value>, String> {
    state.db.with_conn(|conn| {
        let mut stmt = conn.prepare(
            "SELECT id, title, source, chunk_type, chunk_count, created_at FROM knowledge_doc WHERE kb_id = ?1 ORDER BY created_at"
        )?;
        let rows = stmt.query_map(rusqlite::params![kb_id], |row| {
            Ok(serde_json::json!({
                "id": row.get::<_, String>(0)?,
                "title": row.get::<_, Option<String>>(1)?,
                "source": row.get::<_, Option<String>>(2)?,
                "chunkType": row.get::<_, String>(3)?,
                "chunkCount": row.get::<_, i64>(4)?,
                "createdAt": row.get::<_, String>(5)?,
            }))
        })?;
        rows.collect::<Result<Vec<_>, _>>()
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn import_document(
    state: State<'_, AppState>,
    kb_id: String,
    title: String,
    source: String,
    content: String,
    chunk_type: Option<String>,
) -> Result<String, String> {
    let ct = chunk_type.unwrap_or_else(|| "text".to_string());
    crate::rag::indexer::import_document(
        &state.db,
        &*state.vector_store,
        &kb_id,
        &title,
        &source,
        &content,
        &ct,
    )
    .await
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_doc(state: State<'_, AppState>, doc_id: String) -> Result<(), String> {
    crate::rag::indexer::delete_document(&state.db, &*state.vector_store, &doc_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn search_knowledge(
    state: State<'_, AppState>,
    query: String,
    kb_id: Option<String>,
    top_k: Option<usize>,
) -> Result<Vec<serde_json::Value>, String> {
    let results = crate::rag::retriever::search(
        &state.db,
        &*state.vector_store,
        &query,
        kb_id.as_deref(),
        top_k.unwrap_or(5),
    )
    .await
    .map_err(|e| e.to_string())?;

    Ok(results.into_iter().map(|r| serde_json::json!({
        "chunkId": r.chunk_id,
        "text": r.text,
        "score": r.score,
        "docTitle": r.doc_title,
    })).collect())
}
